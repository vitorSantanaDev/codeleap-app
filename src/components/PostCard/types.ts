export type PostCardProps = {
  myPost?: boolean
  post: Post
}

export type Post = {
  id: number
  title: string
  content: string
  username: string
  created_datetime: string
}
