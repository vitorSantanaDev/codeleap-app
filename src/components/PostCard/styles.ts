import styled, { css } from 'styled-components'

import * as HeaderStyles from 'components/Header/styles'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    ${HeaderStyles.Wrapper} {
      border-top-right-radius: ${theme.border.radius};
      border-top-left-radius: ${theme.border.radius};
    }

    border: 1px solid ${theme.colors.gray};
    border-radius: ${theme.border.radius};
    border-radius: ${theme.border.radius};
  `}
`

export const HeaderContent = styled.div`
  ${() => css`
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;
  `}
`

export const ActionButtonsWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    gap: ${theme.spacings.medium};

    img {
      cursor: pointer;
    }
  `}
`

export const Content = styled.div`
  ${({ theme }) => css`
    padding: ${theme.spacings.small};
  `}
`

export const PostInfoWrapper = styled.div`
  ${({ theme }) => css`
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: ${theme.spacings.xsmall};
  `}
`

export const UserNameText = styled.h2`
  ${({ theme }) => css`
    font-weight: ${theme.font.bold};
    color: ${theme.colors.dark_gray};
    font-size: calc(${theme.font.sizes.medium} + 0.2rem);
  `}
`

export const PostTimeText = styled.span`
  ${({ theme }) => css`
    font-weight: ${theme.font.normal};
    color: ${theme.colors.dark_gray};
    font-size: calc(${theme.font.sizes.medium} + 0.2rem);
  `}
`

export const Description = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-weight: ${theme.font.normal};
    font-size: calc(${theme.font.sizes.medium} + 0.2rem);
  `}
`
