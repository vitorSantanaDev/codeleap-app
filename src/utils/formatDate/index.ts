import {
  format,
  parseISO,
  minutesToHours,
  secondsToMinutes,
  differenceInSeconds
} from 'date-fns'

export const formattingStringWithTheTimePassedBetweenTwoDate = (
  hours: number,
  minutes: number,
  seconds: number
): string => {
  if (hours < 1 && minutes < 1) {
    return `${seconds} seconds ago`
  }

  let formattedTime = ''

  if (hours > 0) {
    formattedTime = `${hours} ${hours > 1 ? 'hours' : 'hour'} ago`
  } else if (hours < 1) {
    formattedTime = `${minutes} minutes ago`
  }

  return formattedTime
}

export const rangeOfTimePassedBetweenTwoDates = (
  start: string,
  end: string
): string => {
  const startDate = parseISO(start)
  const endDate = parseISO(end)

  const allSeconds = differenceInSeconds(endDate, startDate)
  const allMinutes = secondsToMinutes(allSeconds)

  const minutesInSeconds = allMinutes * 60
  const seconds = allSeconds - minutesInSeconds

  const hours = minutesToHours(allMinutes)

  const hoursInMinutes = hours * 60

  const minutes = allMinutes - hoursInMinutes

  return hours > 24
    ? format(new Date(startDate), 'dd/MM/yyyy')
    : formattingStringWithTheTimePassedBetweenTwoDate(hours, minutes, seconds)
}
