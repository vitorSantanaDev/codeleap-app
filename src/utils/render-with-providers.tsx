import { Provider } from 'react-redux'
import React, { PropsWithChildren } from 'react'

import theme from 'styles/theme'
import { ThemeProvider } from 'styled-components'
import { RenderOptions, render } from '@testing-library/react'

import { RootState } from 'redux/store'
import { userReducer } from 'redux/actions/user-slice'
import { PreloadedState, configureStore } from '@reduxjs/toolkit'
import { QueryClient, QueryClientProvider } from 'react-query'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  store?: any
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = configureStore({
      reducer: { user: userReducer },
      preloadedState: initialState
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <QueryClientProvider client={new QueryClient()}>
        <Provider store={store}>
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </Provider>
      </QueryClientProvider>
    )
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}
