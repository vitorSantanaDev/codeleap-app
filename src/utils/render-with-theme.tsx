import React, { ReactElement } from 'react'

import theme from 'styles/theme'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'

export const renderWithTheme = (component: ReactElement) =>
  render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
