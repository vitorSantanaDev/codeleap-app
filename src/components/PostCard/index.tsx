import React from 'react'
import { formatISO } from 'date-fns'

import { PostCardProps } from './types'

import Header from 'components/Header'

import EditIcon from 'assets/icons/bx_bx-edit.svg'
import TrashIcon from 'assets/icons/delete-icon.svg'

import { rangeOfTimePassedBetweenTwoDates } from 'utils/formatDate'

import * as S from './styles'

const PostCard: React.FC<PostCardProps> = ({ myPost, post }) => {
  const postTimeLabel = rangeOfTimePassedBetweenTwoDates(
    post.created_datetime,
    formatISO(new Date())
  )

  return (
    <S.Wrapper>
      <Header>
        <S.HeaderContent>
          <span>{post.title}</span>
          {!!myPost && (
            <S.ActionButtonsWrapper>
              <img role="button" src={TrashIcon} alt="Trash icon" />
              <img role="button" src={EditIcon} alt="Edit icon" />
            </S.ActionButtonsWrapper>
          )}
        </S.HeaderContent>
      </Header>
      <S.Content>
        <S.PostInfoWrapper>
          <S.UserNameText>{post.username}</S.UserNameText>
          <S.PostTimeText>{postTimeLabel}</S.PostTimeText>
        </S.PostInfoWrapper>
        <S.Description>{post.content}</S.Description>
      </S.Content>
    </S.Wrapper>
  )
}

export default PostCard
