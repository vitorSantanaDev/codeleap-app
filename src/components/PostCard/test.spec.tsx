import React from 'react'

import PostCard from '.'
import { renderWithTheme } from 'utils/render-with-theme'
import { screen } from '@testing-library/react'

const postMock = {
  id: 20,
  created_datetime: '2023-03-07T00:00:00',
  username: '@Victor',
  content: 'Curabitur suscipit suscipit tellus.',
  title: 'My First Post at CodeLeap Network!'
}

describe('<PostCard />', () => {
  it('should render the post', () => {
    renderWithTheme(<PostCard post={postMock} />)

    const username = screen.getByText(/@victor/i)
    expect(username).toBeInTheDocument()

    const postContent = screen.getByText(/Curabitur suscipit suscipit tellus./i)
    expect(postContent).toBeInTheDocument()

    const postTitle = screen.getByText(/My First Post at CodeLeap Network!/i)
    expect(postTitle).toBeInTheDocument()

    const buttonDeletePost = screen.queryByRole('button', {
      name: /trash icon/i
    })
    expect(buttonDeletePost).not.toBeInTheDocument()

    const buttonEditPost = screen.queryByRole('button', {
      name: /edit icon/i
    })
    expect(buttonEditPost).not.toBeInTheDocument()
  })

  it('should render post with action buttons, when myPost is true', () => {
    renderWithTheme(<PostCard myPost post={postMock} />)

    const buttonDeletePost = screen.getByRole('button', {
      name: /trash icon/i
    })
    expect(buttonDeletePost).toBeInTheDocument()

    const buttonEditPost = screen.getByRole('button', {
      name: /edit icon/i
    })
    expect(buttonEditPost).toBeInTheDocument()
  })

  it('should match snapshot', () => {
    const { container } = renderWithTheme(<PostCard myPost post={postMock} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
