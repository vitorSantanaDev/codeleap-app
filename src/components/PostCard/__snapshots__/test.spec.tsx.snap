// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostCard /> should match snapshot 1`] = `
.c2 {
  font-weight: 700;
  color: #FFFFFF;
  padding: calc(2.4rem + 0.3rem) calc(3.2rem + 0.5rem);
  background-color: #7695EC;
  font-size: calc(2.0rem + 0.2rem);
}

.c0 {
  border: 1px solid #999999;
  border-bottom-right-radius: 0.8rem;
  border-bottom-left-radius: 0.8rem;
}

.c0 .c1 {
  border-top-right-radius: 0.8rem;
  border-top-left-radius: 0.8rem;
}

.c3 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 3.2rem;
}

.c5 {
  padding: 2.4rem;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 1.6rem;
}

.c7 {
  font-weight: 700;
  color: #777777;
  font-size: calc(1.6rem + 0.2rem);
}

.c8 {
  font-weight: 400;
  color: #777777;
  font-size: calc(1.6rem + 0.2rem);
}

.c9 {
  color: #000000;
  font-weight: 400;
  font-size: calc(1.6rem + 0.2rem);
}

<div
  class="c0"
>
  <header
    class="c1 c2"
  >
    <div
      class="c3"
    >
      <span>
        My First Post at CodeLeap Network!
      </span>
      <div
        class="c4"
      >
        <img
          alt="Trash icon"
          role="button"
          src="delete-icon.svg"
        />
        <img
          alt="Edit icon"
          role="button"
          src="bx_bx-edit.svg"
        />
      </div>
    </div>
  </header>
  <div
    class="c5"
  >
    <div
      class="c6"
    >
      <h2
        class="c7"
      >
        @Victor
      </h2>
      <span
        class="c8"
      >
        07/03/2023
      </span>
    </div>
    <p
      class="c9"
    >
      Curabitur suscipit suscipit tellus.
    </p>
  </div>
</div>
`;
